lib/codecs/src/encoding/format/syslog.rs

```diff
@@ -0,0 +1,577 @@
+use bytes::{BufMut, BytesMut};
+use tokio_util::codec::Encoder;
+use vector_core::{config::DataType, event::{Event, LogEvent}, schema};
+use chrono::{DateTime, SecondsFormat, SubsecRound, Utc};
+use vrl::value::{ObjectMap, Value};
+use vector_config::configurable_component;
+
+use std::collections::HashMap;
+
+// All of this block is to support the Facility + Severity enums with convenience of string or ordinal config value:
+use std::str::FromStr;
+use strum::{FromRepr, EnumString, VariantNames};
+// `akin` macro for DRY impl to share with both enums due to lack of a `FromRepr` trait:
+use akin::akin;
+// Custom deserialization with serde needed:
+use serde::{Deserialize, Deserializer, de::Error};
+use serde_aux::field_attributes::deserialize_number_from_string;
+use snafu::{Snafu, OptionExt};
+
+/// Config used to build a `SyslogSerializer`.
+#[configurable_component]
+#[derive(Clone, Debug, Default)]
+#[serde(default)]
+pub struct SyslogSerializerConfig {
+    /// Options for the Syslog serializer.
+    pub syslog: SyslogSerializerOptions
+}
+
+impl SyslogSerializerConfig {
+    /// Build the `SyslogSerializer` from this configuration.
+    pub fn build(&self) -> SyslogSerializer {
+        SyslogSerializer::new(&self)
+    }
+
+    /// The data type of events that are accepted by `SyslogSerializer`.
+    pub fn input_type(&self) -> DataType {
+        DataType::Log
+    }
+
+    /// The schema required by the serializer.
+    pub fn schema_requirement(&self) -> schema::Requirement {
+        schema::Requirement::empty()
+    }
+}
+
+/// Syslog serializer options.
+#[configurable_component]
+#[derive(Clone, Debug, Default)]
+// Serde default makes all config keys optional.
+// Each field assigns either a fixed value, or field name (lookup field key to retrieve dynamic value per `LogEvent`).
+#[serde(default)]
+pub struct SyslogSerializerOptions {
+    /// RFC
+    rfc: SyslogRFC,
+    /// Facility
+    #[serde(deserialize_with = "Facility::deserialize")]
+    facility: Facility,
+    /// Severity
+    #[serde(deserialize_with = "Severity::deserialize")]
+    severity: Severity,
+
+    /// App Name
+    app_name: Option<String>,
+    /// Proc ID
+    proc_id: Option<String>,
+    /// Msg ID
+    msg_id: Option<String>,
+
+    /// Payload key
+    payload_key: String,
+
+    // Q: The majority of the fields above pragmatically only make sense as config for keys to query?
+}
```

Comment:
This was adapted from the original PR, `payload_key` has since been removed in favor of the actual Vector log `message` when available or could be set via VRL.

In my local branch I was looking into having most of these field that have an equivalent syslog struct field to instead share the same struct:

```rust
pub struct SyslogSerializerOptions {
    /// RFC
    rfc: SyslogRFC,

    #[serde(flatten)]
    #[configurable(derived)]
    priority: Pri,

    #[serde(flatten)]
    #[configurable(derived)]
    tag: Tag,
}
```

There was another variation that instead of configuring overrides to the defaults, would use `ConfigTargetPath` as a more appropriate config approach by allowing the user to adjust the VRL keys to map into the syslog struct fields:

```rust
pub struct SyslogSerializerOptions {
    /// Facility
    #[default = "facility"]
    facility: ConfigTargetPath,
    /// Severity
    #[default = "severity"]
    severity: ConfigTargetPath,

    /// Application Name
    #[default = "appname"]
    app_name: ConfigTargetPath,
    /// Process ID
    #[default = "procid"]
    proc_id: ConfigTargetPath,
    /// Message ID
    #[default = "msgid"]
    msg_id: ConfigTargetPath,

    /// Hostname
    #[default = "hostname"]
    hostname: ConfigTargetPath,
    /// Message
    message: Option<ConfigTargetPath>,
    /// Timestamp
    timestamp: Option<ConfigTargetPath>,
    /// Structured Data
    structured_data: Option<ConfigTargetPath>,
}
```

Probably the right way to go?